File: 백준/Silver/11403. 경로 찾기/경로 찾기.py
N = int(input())

G = [[] for _ in range(N)]
pan = [list(map(int, input().split())) for _ in range(N)]

for i in range(N):
    for j in range(N):
        if i == j:
            continue
        if pan[i][j]:
            G[i].append(j)


def dfs(first, x, y, depth):
    global flag
    if x == y and depth:
        flag = 1
        return
    if first != x:
        visited[x] = 1
    for w in G[x]:
        if not visited[w]:
            dfs(first, w, y, depth+1)


answer = [[0 for _ in range(N)] for _ in range(N)]
for i in range(N):
    for j in range(N):
        flag = 0
        visited = [0 for _ in range(N)]
        dfs(i, i, j, 0)
        if flag:
            answer[i][j] = 1

for v in answer:
    print(*v)

File: 백준/Silver/11403. 경로 찾기/경로 찾기.js
const fs = require("fs");
const stdin = (
  process.platform === "linux"
    ? fs.readFileSync("/dev/stdin").toString()
    : `7
    0 0 0 1 0 0 0
    0 0 0 0 0 0 1
    0 0 0 0 0 0 0
    0 0 0 0 1 1 0
    1 0 0 0 0 0 0
    0 0 0 0 0 0 1
    0 0 1 0 0 0 0
`
)
  .trim()
  .split("\n");

const input = (() => {
  let line = 0;
  return () => stdin[line++];
})();

const N = parseInt(input().trim());
const G = Array.from(Array(N), () => Array());

for (let k = 0; k < N; k++) {
  const inputArray = input().trim().split(" ").map(Number);
  for (let p = 0; p < N; p++) {
    if (inputArray[p] === 1) {
      G[k].push(p);
    }
  }
}

const answer = Array.from(Array(N), () => Array(N).fill(0));

for (let i = 0; i < N; i++) {
  for (let j = 0; j < N; j++) {
    const visited = Array(N).fill(0);
    dfs(i, i, j, 0, visited);
  }
}

answer.forEach((v) => {
  console.log(v.join(" "));
});

function dfs(first, x, y, depth, visited) {
  if (x === y && depth > 0) {
    answer[first][y] = 1;
    return;
  }

  if (first !== x) {
    visited[x] = 1;
  }

  G[x].forEach((w) => {
    if (!visited[w]) {
      dfs(first, w, y, depth + 1, visited);
    }
  });
}
